src/env/double_write.rs

```diff
@@ -0,0 +1,722 @@
+            }
+            std::cmp::Ordering::Less => {
+                self.fd2.read().unwrap().as_ref().unwrap().read(offset, buf)
+            }
+        }
+    }
+
+    fn write(&self, offset: usize, content: &[u8]) -> IoResult<usize> {
+        block_on(self.wait_one(FileTask::Write {
+            offset,
+            bytes: content.to_vec(),
+        }))
+        .map(|x| x.unwrap())
+    }
+
+    fn allocate(&self, offset: usize, size: usize) -> IoResult<()> {
+        block_on(self.wait_one(FileTask::Allocate { offset, size })).map(|_| ())
+    }
+
+    async fn wait_one(&self, task: FileTask) -> IoResult<Option<usize>> {
+        let (cb1, mut f1) = paired_future_callback();
+        let (cb2, mut f2) = paired_future_callback();
+        self.disk1.send((task.clone(), cb1)).unwrap();
+        self.disk2.send((task, cb2)).unwrap();
+
+        select! {
+            res1 = f1 => res1.unwrap(),
+            res2 = f2 => res2.unwrap(),
+        }
+    }
+}
+
+impl<F: FileSystem> Drop for HedgedHandle<F> {
+    fn drop(&mut self) {
+        self.disk1.send((FileTask::Stop, Box::new(|_| {}))).unwrap();
+        self.disk2.send((FileTask::Stop, Box::new(|_| {}))).unwrap();
+        self.handle1.take().unwrap().join().unwrap();
+        self.handle2.take().unwrap().join().unwrap();
+    }
+}
+
+impl<F: FileSystem> Handle for HedgedHandle<F> {
+    fn truncate(&self, offset: usize) -> IoResult<()> {
+        block_on(self.wait_one(FileTask::Truncate(offset))).map(|_| ())
+    }
+
+    fn file_size(&self) -> IoResult<usize> {
+        block_on(self.wait_one(FileTask::FileSize)).map(|x| x.unwrap())
+    }
+
+    fn sync(&self) -> IoResult<()> {
+        block_on(self.wait_one(FileTask::Sync)).map(|_| ())
+    }
+}
+
+pub struct HedgedWriter<F: FileSystem> {
+    disk1: Sender<(FileTask, Callback<usize>)>,
+    disk2: Sender<(FileTask, Callback<usize>)>,
+    writer1: Arc<F::Writer>,
+    writer2: Arc<F::Writer>,
+}
+
+impl<F: FileSystem> HedgedWriter<F> {
+    pub fn new(disk1: Sender<(FileTask, Callback<usize>)>, disk2: Sender<(FileTask, Callback<usize>)>, writer1: Arc<F::Writer>, writer2: Arc<F::Writer>) -> Self {
+      
+        Self {
+            handle,
+
+        }
+    }
+}
+
+impl<F: FileSystem> Write for HedgedWriter<F> {
+    fn write(&mut self, buf: &[u8]) -> IoResult<usize> {
+    }
+
+    fn flush(&mut self) -> IoResult<()> {
+    }
+}
+
+impl<F: FileSystem> WriteExt for HedgedWriter<F> {
+    fn truncate(&mut self, offset: usize) -> IoResult<()> {
+    }
+
+    fn allocate(&mut self, offset: usize, size: usize) -> IoResult<()> {
+    }
+}
+
+impl<F: FileSystem> Seek for HedgedWriter<F> {
+    fn seek(&mut self, pos: SeekFrom) -> IoResult<u64> {
+        match pos {
+            SeekFrom::Start(offset) => self.offset = offset as usize,
+            SeekFrom::Current(i) => self.offset = (self.offset as i64 + i) as usize,
+            SeekFrom::End(i) => self.offset = (self.inner.file_size()? as i64 + i) as usize,
```

Comment:
The file size on the two disks can be different, right?  Looks like it will cause problems

Reply:
It would always return the file size of the newer disk. The result on each call for a being appended file must be monotonic increasing. lt has nothing different compared to the non-hedging file system.